### ISSUE TITLE
Updating multiple instances of a cluster sharing one root directory

### ISSUE THREAD

**Issue by TheRealestMVP** [FezVrasta/ark-server-tools#796]
TL;DR at the bottom.

I'm running 4 instances on the same server using arkmanager. Those are different maps, all clustered together.

## Old situation

I previously had 4 separate installation directories, and I was updating all 4 of them separately via cron (`arkmanager update --update-mods --safe --warn @instanceName` for all 4 of them). This worked fine. But it has some issues: It takes 4 times the disk space, INI files must be maintained 4 times, backup must be managed 4 times, etc.. Also, arkmanager seems to download the complete mods every time it checks for updates and for each instance. Since I check for updates every 3 hours and have a total mod size of about 1 GB this adds up to 16 GB traffic per day(!), even when no updates are available.

## New situation

I now switched to a single installation directory. The instances share the same `arkserverroot`, the same set of mods, and of course `arkopt_ClusterDirOverride`, and have individual `ark_AltSaveDirectoryName` settings.

Example island.cfg (excerpt):
```
arkserverroot=/home/steam/ark-server
serverMap=TheIsland
ark_AltSaveDirectoryName=SavedArks-Island
arkopt_clusterid=mycluster
arkopt_ClusterDirOverride=/home/steam/ark-cluster-data
```

Example scorched.cfg (excerpt):
```
arkserverroot=/home/steam/ark-server
serverMap=ScorchedEarth_P
ark_AltSaveDirectoryName=SavedArks-Scorched
arkopt_clusterid=mycluster
arkopt_ClusterDirOverride=/home/steam/ark-cluster-data
```

Same for the other instances (you get the idea).

So the save files are separated into different directories within the `/home/steam/ark-server/ShooterGame/Saved/` directory, and all instances use the same Game.ini and GameUserSettings.ini files, which is what I want.

The whole setup works pretty well so far.

## The problem

The problem now is that automatic updating and restarting does no longer work for all the instances. The game and the mods in the shared root directory are updated, but only the first instance is restarted.

When I run `arkmanager update --update-mods --safe --warn @all` and updates are available, arkmanager installs them when processing the first instance (island.cfg in my case, may be random), then restarts this instance. All subsequent instances then do not detect updates and keep running the old version, even though game/mods have been updated due to the same root directory.

The same goes with running `arkmanager update --update-mods --safe --warn @instanceName` individually. Only the first one is restarted, the others do not detect the update.

From various server guides on the net I learned that the shared root directory for multiple server instances is a supported setup. Does arkmanager not support this? Am I doing this wrong?

If it's not possible at the current state, I would like to make the following suggestion: Add an option to arkmanager to force a restart of all the given instances if any of them has been updated.

**TL;DR:** Updating multiple instances with the same shared `arkserverroot` only restarts the first instance. The other running instances do not re-detect the updates and are not restarted.

**Comment by jtmalos** [comment]
I've had this problem for a long time and have resolved it with additional scripts calling ark-tools. 

Essentially each server that is running in parallel with the one being updated retains the deprecated version in memory and needs a restart to load the now-updated files common to the cluster. Instead of all servers being on the same revision, we were getting them staggered in revision and desynchronising. This was also due to our policy to not update/restart servers if players were present, so depending on who was on when, some servers were never getting the restart and lagging behind by multiple revisions. 

In essence, our script is called regularly by cron (every 30mins or so) and does the following:
- check for an update on the first available (non-occupied) server (noplayers)
- if an update occurs, set a flag (touch a set of files called "server?_needs_restart") for each server
- immediately clear the flag for the server being updated
- after the update call to ark-tools is completed, loop through each of the flags
- if set, clear the flag and call server restart

My scripts are very hairy and not fit for pubic consumption, but they perform additional finesse such as to first perform a download and set a flag before attempting updates, and also to perform separate updates for mods since these checks take much longer (essentially re-downloading all mods). It's quite likely aspects are now redundant since I haven't been following all of the new features of ark-tools. But since it addresses this thread's problem and my scripts aren't broken ...

Hope that helps.

**Comment by meswrex** [comment]
Heh; take 2; wrong account :)

I've had this problem for a long time and have resolved it with additional scripts calling ark-tools.

Essentially each server that is running in parallel with the one being updated retains the deprecated version in memory and needs a restart to load the now-updated files common to the cluster. Instead of all servers being on the same revision, we were getting them staggered in revision and desynchronising. This was also due to our policy to not update/restart servers if players were present, so depending on who was on when, some servers were never getting the restart and lagging behind by multiple revisions.

In essence, our script is called regularly by cron (every 30mins or so) and does the following:

check for an update on the first available (non-occupied) server (noplayers)
if an update occurs, set a flag (touch a set of files called "server?_needs_restart") for each server
immediately clear the flag for the server being updated
after the update call to ark-tools is completed, loop through each of the flags
if set, clear the flag and call server restart
My scripts are very hairy and not fit for pubic consumption, but they perform additional finesse such as to first perform a download and set a flag before attempting updates, and also to perform separate updates for mods since these checks take much longer (essentially re-downloading all mods). It's quite likely aspects are now redundant since I haven't been following all of the new features of ark-tools. But since it addresses this thread's problem and my scripts aren't broken ...

Hope that helps.

**Comment by meswrex** [comment]
If you are happy to force a restart with @all then there's no problem. But if you want to wait until a server is vacant for player convenience, then I'm not sure if using @all will remember that a server wasn't restarted due to being occupied. 
The advantage of setting file flags is that they stay set until the server is eventually restarted. This allows for waiting until a given server is vacant. If it isn't and you don't succeed in restarting due to a noplayers switch, then the flag will remain set until the next cronjob and the check for restart will be performed again. Eventually all servers will be reset when they become vacant and will continue checking until the flag is cleared.

**Comment by baeri** [comment]
I made a quick'n'dirty PR. It restarts all instances if the first needs an update.

This only works, if you use `update [...] @all` as paramter. Using `update [...] @instance1` and `update [...] @instance2` after this won't work yet.

**Comment by TheRealestMVP** [comment]
> I made a quick'n'dirty PR. It restarts all instances if the first needs an update.

My shell scripting is quite noobish, but wouldn't this detect game updates only and fail on mod updates?

Honestly I tried something similar with a wrapper script. Script checks the current game build number by parsing `arkmanager status @firstInstance`, then attempts the update on that first instance and checks the build number again. If changed, restarts all other instances afterwards. Felt really wonky and worked 2 days until only a mod was updated.

**Comment by klightspeed** [comment]
Does 734e71d fix this?

**Comment by baeri** [comment]
@klightspeed, yeah, this works.
